src/flint/types/fmpz_mod.pyx

```diff
@@ -0,0 +1,428 @@
+        return res == 1
+
+    def __richcmp__(s, t, int op):
+        cdef bint res
+        if op != 2 and op != 3:
+            raise TypeError("fmpz_mod cannot be ordered")
+        # TODO: is this the best method for comparison?
+        if typecheck(s, fmpz_mod) and typecheck(t, fmpz_mod):
+            res = ((<fmpz_mod>s).val[0] == (<fmpz_mod>t).val[0]) and \
+                  ((<fmpz_mod>s).ctx.val.n == (<fmpz_mod>t).ctx.val.n)
+            if op == 2:
+                return res
+            else:
+                return not res
+        # TODO: is this the best method for comparison?
+        # Seems like I'm doing too many type conversions?
+        elif typecheck(s, fmpz_mod) and typecheck(t, int):
+            res = int(s) == t % (<fmpz_mod>s).ctx.modulus()
+            if op == 2:
+                return res
+            else:
+                return not res
+        return NotImplemented
+
+    def __bool__(self):
+        return not self.is_zero()
+
+    def __repr__(self):
+        return "fmpz_mod({}, {})".format(
+            fmpz_get_intlong(self.val),
+            self.ctx.modulus()
+        )
+
+    # TODO: seems ugly...
+    def __hash__(self):
+        return  hash(
+            (int(self), int(self.ctx.modulus()))
+        )
+
+    def __int__(self):
+        return fmpz_get_intlong(self.val)
+
+    def __str__(self):
+        return str(int(self))
+
+    # ---------------- #
+    #    Arithmetic    #
+    # ---------------- #
+
+    def __pos__(self):
+        return self
+
+    def __neg__(self):
+        cdef fmpz_mod res
+        res = fmpz_mod.__new__(fmpz_mod)
+        res.ctx = self.ctx
+
+        fmpz_mod_neg(
+            res.val, self.val, 
+            (<fmpz_mod_ctx_t>self.ctx.val))
+        return res
+
+    def __add__(self, other):
+        cdef fmpz_mod res
+        res = fmpz_mod.__new__(fmpz_mod)
+        res.ctx = self.ctx
+
+        # Add two fmpz_mod if moduli match
+        if typecheck(other, fmpz_mod):
+            if not self.ctx.val.n == (<fmpz_mod>other).ctx.val.n:
+                raise ValueError("moduli must match")
+
+            fmpz_mod_add(
+                res.val, self.val, (<fmpz_mod>other).val, self.ctx.val
+            )
+            return res
+
+        # Attempt to add an fmpz to an fmpz_mod
+        other = any_as_fmpz(other)
+        if other is NotImplemented:
+            raise NotImplementedError
+
+        fmpz_mod_add_fmpz(
+            res.val, self.val, (<fmpz>other).val, self.ctx.val
+        )
+
+        return res
+
+    def __radd__(self, other):
+        return self.__add__(other)
+
+    def __iadd__(self, other):
+        # Add two fmpz_mod if moduli match
+        if typecheck(other, fmpz_mod):
+            if not self.ctx.val.n == (<fmpz_mod>other).ctx.val.n:
+                raise ValueError("moduli must match")
+
+            fmpz_mod_add(
+                self.val, self.val, (<fmpz_mod>other).val, self.ctx.val
+            )
+            return self
```

Comment:
Generally in Python something like this would not have an in-place mutating method e.g. `fmpz` does not, `int` does not etc. Mutable container types like a matrix would but I would not have defined any in-place methods for `fmpz_mod`.

Does it really make sense to add this (and other mutating methods)?

Reply:
If `fmpz` doesn't then I suppose we can remove these. The only reason I added them is it should be slightly faster as you don't have to init the result? Not sure if that level of benchmarking is interesting?

Reply:
I would be interested to hear benchmarks but regardless the divergence in behaviour from all other numeric types in Python would just confuse people and lead to bugs. Understanding exactly how `+=` works with mutable types is something that confuses a lot of people especially when passing values to functions:
```python
def f(a):
    a += 1
    return a**2

b = 2
c = f(b)
# Most people would be suprised if this did not print 2:
print(b)
```

Reply:
Very naive benchmark, but can do something more scientific later

```py
>>> from flint import *
>>> import timeit
>>>
>>> Fp = fmpz_mod_ctx(10934940634374701148446891134773061114063995215204380092061981208900044999798983348516563438926479516948194331363194666667838046483659456628774161285399003)
>>>  
>>> a = Fp(1312529757321787310305008287695652105910635919800619521256615619505416908854385534618644512159270554422685672404970180005984257654287906025857278585166015)
>>> b = Fp(2743993601401652712330999846100462585643293332719558784670795396207932456640420188396741854989884607067889476240198970284190649813933943542680798609562697)
>>> 
>>> timeit.timeit("c = a + b; c += b", globals={"a":a, "b":b}, number=1_000_000)
0.2847036040038802
>>> timeit.timeit("c = a + b; c = c + b", globals={"a":a, "b":b}, number=1_000_000)
0.4899688330187928
```

Reply:
I think that a lower-level API could allow mutation but we should not allow it for the `fmpz_mod` type that is added here.